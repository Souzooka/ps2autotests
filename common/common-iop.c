#include <stdio.h>
#include <stdarg.h>
#include <xprintf.h>

#define SCHEDF_BUFFER_SIZE 65536

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsnprintf(schedfBuffer + schedfBufferPos, SCHEDF_BUFFER_SIZE, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
