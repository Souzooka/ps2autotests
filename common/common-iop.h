#ifndef PS2AUTOTESTS_COMMON_IOP_H
#define PS2AUTOTESTS_COMMON_IOP_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _IOP
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

void schedf(const char *format, ...);
void flushschedf();

#endif
