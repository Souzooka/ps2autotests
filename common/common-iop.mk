# _____     ___ ____     ___ ____
#  ____|   |    ____|   |        | |____|
# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
#-----------------------------------------------------------------------
# Copyright 2001-2004, ps2dev - http://www.ps2dev.org
# Licenced under Academic Free License version 2.0
# Review ps2sdk README & LICENSE files for further details.
#
# $Id: Makefile.eeglobal_cpp_sample 1525 2009-01-23 00:24:13Z radad $

# Define pefixes for the various toolchain binaries.
include $(COMMON_DIR)/common-defines.mk

ASFLAGS_TARGET = -mcpu=r3000

CFLAGS_TARGET  = -miop
ASFLAGS_TARGET = -march=r3000
LDFLAGS_TARGET = -miop

# include dir
IOP_INCS := -I$(PS2SDK)/iop/include -I$(PS2SDK)/common/include -I$(COMMON_DIR) \
	-I. $(IOP_INCS)
# C compiler flags
IOP_CFLAGS := $(CFLAGS_TARGET) -O2 -G0 $(IOP_INCS) $(IOP_CFLAGS)

# linker flags
IOP_LDFLAGS := $(LDFLAGS_TARGET) -nostdlib -L$(PS2SDK)/iop/lib $(IOP_LDFLAGS)

# asssembler flags
IOP_ASFLAGS := $(ASFLAGS_TARGET) -EL -G0 $(IOP_ASFLAGS)

# link with following libraries (libs need to be defined multiple times in order for linking to work!!)
IOP_LIBS += -lkernel -lgcc

# Externally defined variables: IOP_BIN, IOP_OBJS, IOP_LIB

%.o : %.c
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.S
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.s
	$(IOP_AS) $(IOP_ASFLAGS) $< -o $@

%.o : %.lst
	echo "#include \"irx_imports.h\"" > build-imports.c
	cat $< >> build-imports.c
	$(IOP_CC) $(IOP_CFLAGS) -c build-imports.c -o $@
	rm -f build-imports.c

%.irx : %.o $(EXTRA_OBJS)
	$(IOP_CC) $(IOP_LDFLAGS) -o $@ $< $(EXTRA_OBJS) $(IOP_LIBS)

all: $(TARGETS:=.irx)

clean:
	-$(RM) -f $(TARGETS:=.irx) $(TARGETS:=.o)

rebuild: clean all
