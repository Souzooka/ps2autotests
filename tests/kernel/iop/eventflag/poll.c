#include <common-iop.h>
#include <kerr.h>
#include <thevent.h>
#include <thbase.h>

#define STACK_SIZE 0x800

void waitThreadProc(u32 eventFlagId) {
	WaitEventFlag(eventFlagId, ~0, WEF_OR | WEF_CLEAR, NULL);
}

int createTestThread(void *entry, int prio, u32 stackSize) {
	iop_thread_t threadParam;
	memset(&threadParam, 0, sizeof(iop_thread_t));
	threadParam.attr      = TH_C;
	threadParam.thread    = entry;
	threadParam.priority  = prio;
	threadParam.stacksize = stackSize;
	return CreateThread(&threadParam);
}

int getThreadPriority(int threadId) {
	iop_thread_info_t threadStat;
	memset(&threadStat, 0, sizeof(iop_thread_info_t));
	int result = ReferThreadStatus(threadId, &threadStat);
	if (result >= 0) {
		return threadStat.currentPriority;
	} else {
		return -1;
	}
}

int createTestEventFlag(u32 attr, u32 option, u32 initPattern) {
	iop_event_t eventFlagParams;
	memset(&eventFlagParams, 0, sizeof(eventFlagParams));
	eventFlagParams.attr   = attr;
	eventFlagParams.option = option;
	eventFlagParams.bits   = initPattern;
	return CreateEventFlag(&eventFlagParams);
}

void doPollEventFlag(int eventFlagId, u32 waitPattern, int mode, u32 *resultPattern) {
	int result = PollEventFlag(eventFlagId, waitPattern, mode, resultPattern);
	if(result == KE_OK) {
		schedf("succeeded");
	} else {
		schedf("failed -> result: %d", result);
	}
	if(resultPattern != NULL) {
		schedf(", resultPattern: %08x", *resultPattern);
	}
	schedf("\n");
}

void testPollBusyEventFlag(u32 eventFlagMode) {
	int eventFlagId = createTestEventFlag(0, eventFlagMode, 0);
	u32 resultPattern = 0xCCCC;
	
	int currentThreadPriority = getThreadPriority(TH_SELF);
	int threadId = createTestThread((void*)&waitThreadProc, currentThreadPriority - 1, STACK_SIZE);
	StartThread(threadId, (void*)eventFlagId);
	
	doPollEventFlag(eventFlagId, 0x33, WEF_AND, &resultPattern);
	
	TerminateThread(threadId);
	DeleteThread(threadId);
	
	DeleteEventFlag(eventFlagId);
}

int _start(int argc, char **argv) {
	printf("-- TEST BEGIN\n");
	
	{
		schedf("match success cases:\n");
		
		int eventFlagId = createTestEventFlag(EA_SINGLE, 0, 0xF);
		u32 resultPattern = 0xCCCC;
		
		schedf("  poll with matching pattern, OR wait mode: ");
		doPollEventFlag(eventFlagId, 0x1, WEF_OR, &resultPattern);
		
		schedf("  poll with matching pattern, AND wait mode: ");
		doPollEventFlag(eventFlagId, 0xF, WEF_AND, &resultPattern);
		
		DeleteEventFlag(eventFlagId);
		flushschedf();
	}
	
	{
		schedf("match failure cases:\n");
		
		int eventFlagId = createTestEventFlag(EA_SINGLE, 0, 0xF);
		u32 resultPattern = 0xCCCC;
		
		schedf("  poll with non-matching pattern, OR wait mode: ");
		doPollEventFlag(eventFlagId, 0x10, WEF_OR, &resultPattern);
		
		schedf("  poll with non-matching pattern, AND wait mode: ");
		doPollEventFlag(eventFlagId, 0x1F, WEF_AND, &resultPattern);
		
		DeleteEventFlag(eventFlagId);
		flushschedf();
	}
	
	{
		schedf("clear on success:\n");
		
		int eventFlagId = createTestEventFlag(EA_SINGLE, 0, 0x3333);
		u32 resultPattern = 0xCCCC;
		
		schedf("  poll with matching pattern, OR and CLEAR wait mode: ");
		doPollEventFlag(eventFlagId, 0x3, WEF_OR | WEF_CLEAR, &resultPattern);
		
		schedf("  poll a second time with matching pattern, OR wait mode: ");
		resultPattern = 0xCCCC;
		doPollEventFlag(eventFlagId, 0x30, WEF_OR, &resultPattern);
		
		DeleteEventFlag(eventFlagId);
		flushschedf();
	}
	
	{
		schedf("test busy event flag:\n");
		
		schedf("  single wait event flag: ");
		testPollBusyEventFlag(EA_SINGLE);
		
		schedf("  multi wait event flag: ");
		testPollBusyEventFlag(EA_MULTI);
	}
	
	{
		schedf("corner cases:\n");
		
		int eventFlagId = createTestEventFlag(EA_SINGLE, 0, ~0);
		u32 resultPattern = 0xCCCC;
		
		schedf("  poll with zero pattern: ");
		doPollEventFlag(eventFlagId, 0, WEF_AND, &resultPattern);
		
		schedf("  poll with null result pattern: ");
		doPollEventFlag(eventFlagId, 0x1, WEF_AND, NULL);
		
		DeleteEventFlag(eventFlagId);
		
		schedf("  poll with invalid tid: ");
		doPollEventFlag(0, ~0, 0, NULL);
		
		flushschedf();
	}
	
	printf("-- TEST END\n");
	
	return 0;
}
