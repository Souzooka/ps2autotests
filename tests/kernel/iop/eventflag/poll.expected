-- TEST BEGIN
match success cases:
  poll with matching pattern, OR wait mode: succeeded, resultPattern: 0000000f
  poll with matching pattern, AND wait mode: succeeded, resultPattern: 0000000f
match failure cases:
  poll with non-matching pattern, OR wait mode: failed -> result: -421, resultPattern: 0000cccc
  poll with non-matching pattern, AND wait mode: failed -> result: -421, resultPattern: 0000cccc
clear on success:
  poll with matching pattern, OR and CLEAR wait mode: succeeded, resultPattern: 00003333
  poll a second time with matching pattern, OR wait mode: failed -> result: -421, resultPattern: 0000cccc
test busy event flag:
  single wait event flag: failed -> result: -422, resultPattern: 0000cccc
  multi wait event flag: failed -> result: -422, resultPattern: 0000cccc
corner cases:
  poll with zero pattern: failed -> result: -423, resultPattern: 0000cccc
  poll with null result pattern: succeeded
  poll with invalid tid: failed -> result: -409
-- TEST END
